File: eosensor.py
import sys
import clr
import System.Collections.Generic
import System
clr.AddReference('System.Core')
clr.AddReference('IronPython')
clr.AddReference('System.Xml')
clr.AddReferenceByName('Utilities')
clr.AddReferenceByName('HSFUniverse')
clr.AddReferenceByName('UserModel')
clr.AddReferenceByName('MissionElements')
clr.AddReferenceByName('HSFSystem')
#clr.AddReferenceByName('SystemState')
import System.Xml
import HSFSystem
import HSFSubsystem
import MissionElements
import Utilities
import HSFUniverse
import UserModel
# from MissionElements import SystemState
from HSFSystem import *
from System.Xml import XmlNode
from Utilities import *
from HSFUniverse import *
from UserModel import *
from MissionElements import *
from System import Func, Delegate, Math
from System.Collections.Generic import Dictionary, KeyValuePair
from IronPython.Compiler import CallTarget0

class eosensor(HSFSubsystem.Subsystem):
    def __new__(cls, node, asset):
        instance = HSFSubsystem.Subsystem.__new__(cls)
        instance.Asset = asset
        instance.Name = instance.Asset.Name + '.' + node.Attributes['subsystemName'].Value.ToString().ToLower()

        instance.PIXELS_KEY = Utilities.StateVariableKey[System.Double](instance.Asset.Name + '.' + 'numpixels')
        instance.INCIDENCE_KEY = Utilities.StateVariableKey[System.Double](instance.Asset.Name + '.' + 'incidenceangle')
        instance.EOON_KEY = Utilities.StateVariableKey[System.Boolean](instance.Asset.Name + '.' + 'eosensoron')
        instance.addKey(instance.PIXELS_KEY)
        instance.addKey(instance.INCIDENCE_KEY)
        instance.addKey(instance.EOON_KEY)

        instance._lowQualityPixels = 5000
        instance._lowQualityTime = 3
        instance._midQualityPixels = 10000
        instance._midQualityTime = 5
        instance._highQualityPixels = 15000
        instance._highQualityTime = 7
        if (node.Attributes['lowQualityPixels'] != None):
            instance._lowQualityPixels = float(node.Attributes['lowQualityPixels'].Value.ToString())
        if (node.Attributes['lowQualityTime'] != None):
            instance._lowQualityTime = float(node.Attributes['lowQualityTime'].Value.ToString())
        if (node.Attributes['midQualityPixels'] != None):
            instance._midQualityPixels = float(node.Attributes['midQualityPixels'].Value.ToString())
        if (node.Attributes['midQualityTime'] != None):
            instance._midQualityTime = float(node.Attributes['midQualityTime'].Value.ToString())
        if (node.Attributes['highQualityPixels'] != None):
            instance._highQualityPixels = float(node.Attributes['highQualityPixels'].Value.ToString())
        if (node.Attributes['highQualityTime'] != None):
            instance._highQualityTime = float(node.Attributes['highQualityTime'].Value.ToString())

        return instance

    def CanPerform(self, event, universe):
         if (self._task.Type == "imaging"):
             value = self._task.Target.Value
             pixels = self._lowQualityPixels
             timetocapture = self._lowQualityTime
             if (value <= self._highQualityTime and value >= self._midQualityTime):
                    pixels = self._midQualityPixels
                    timetocapture = self._midQualityTime
             if (value > self._highQualityTime):
                pixels = self._highQualityPixels
                timetocapture = self._highQualityTime

             es = event.GetEventStart(self.Asset)
             ts = event.GetTaskStart(self.Asset)
             te = event.GetTaskEnd(self.Asset)
             if (ts > te):
                # Logger.Report("EOSensor lost access")
                 return False
             te = ts + timetocapture
             event.SetTaskEnd(self.Asset, te)

             position = self.Asset.AssetDynamicState
             timage = ts + timetocapture / 2
             m_SC_pos_at_tf_ECI = position.PositionECI(timage)
             m_target_pos_at_tf_ECI = self._task.Target.DynamicState.PositionECI(timage)
             m_pv = m_target_pos_at_tf_ECI - m_SC_pos_at_tf_ECI
             pos_norm = -m_SC_pos_at_tf_ECI / Matrix[System.Double].Norm(-m_SC_pos_at_tf_ECI)
             pv_norm = m_pv / Matrix[System.Double].Norm(m_pv)

             incidenceang = 90 - 180 / Math.PI * Math.Acos(Matrix[System.Double].Dot(pos_norm, pv_norm))
             self._newState.AddValue(self.INCIDENCE_KEY, timage, incidenceang)
             self._newState.AddValue(self.INCIDENCE_KEY, timage + 1, 0.0)
             self._newState.AddValue(self.PIXELS_KEY,timage, pixels)
             self._newState.AddValue(self.PIXELS_KEY, timage + 1, 0.0)
             self._newState.AddValue(self.EOON_KEY, ts, True)
             self._newState.AddValue(self.EOON_KEY, te, False)
             #print("EOSensor CanPreform, Imaging Task, IncidenceAng")
             #print(incidenceang)
             #print("EOSensor CanPreform, Imaging Task, Pixels")
             #print(pixels)
             return True

    def CanExtend(self, event, universe, extendTo):
        return super(eosensor, self).CanExtend(event, universe, extendTo)

    def Power_asset1_from_EOSensor_asset1(self, event):
        prof1 = HSFProfile[System.Double]()
        prof1[event.GetEventStart(self.Asset)] = 10
        if (event.State.GetValueAtTime(self.EOON_KEY, event.GetTaskStart(self.Asset)).Item2):
            prof1[event.GetTaskStart(self.Asset)] = 60
            prof1[event.GetTaskEnd(self.Asset)] = 10
        return prof1

    def SSDR_asset1_from_EOSensor_asset1(self, event):
        return event.State.GetProfile(self.PIXELS_KEY) / 500

    def DepFinder(self, depFnName):  # Search for method from string input
        fnc = getattr(self, depFnName)
        dep = Dictionary[str, Delegate]()
        depFnToAdd = Func[Event, Utilities.HSFProfile[System.Double]](fnc)
        dep.Add(depFnName, depFnToAdd)
        return dep

    def GetDependencyCollector(self):
        return Func[Event,  Utilities.HSFProfile[System.Double]](self.DependencyCollector)

    def DependencyCollector(self, currentEvent):
        return super(eosensor, self).DependencyCollector(currentEvent)

#    def GetDependencyDictionary(self):
#        dep = Dictionary[str, Delegate]()
#        depFunc1 = Func[Event,  Utilities.HSFProfile[System.Double]](self.POWERSUB_PowerProfile_EOSENSORSUB)
#        dep.Add("PowerfromEOSensor" + "." + self.Asset.Name, depFunc1)
#        depFunc1 = Func[Event,  Utilities.HSFProfile[System.Double]](self.SSDRSUB_NewDataProfile_EOSENSORSUB)
#        dep.Add("SSDRfromEOSensor" + "." + self.Asset.Name, depFunc1)
#        return dep

File: HSFSystem/EOSensor.cs
// Copyright (c) 2016 California Polytechnic State University
// Authors: Morgan Yost (morgan.yost125@gmail.com) Eric A. Mehiel (emehiel@calpoly.edu)

using System;
using System.Collections.Generic;
using HSFUniverse;
using MissionElements;
using System.Xml;
using HSFSystem;
using Utilities;
using System.Diagnostics.CodeAnalysis;

namespace HSFSubsystem
{
    //[ExcludeFromCodeCoverage]
    public class EOSensor : Subsystem
    {
        #region Attributes
        //Default Values
        public static string SUBNAME_EOSENSOR = "EOSensor";
        protected StateVariableKey<double> PIXELS_KEY;
        protected StateVariableKey<double> INCIDENCE_KEY;
        protected StateVariableKey<bool> EOON_KEY;
        protected double _lowQualityPixels = 5000;
        protected double _lowQualityCaptureTime = 3;
        protected double _midQualityPixels = 10000;
        protected double _midQualityCaptureTime = 5;
        protected double _highQualityPixels = 15000;
        protected double _highQualityCaptureTime = 7;
        #endregion

        #region Constructors
        /// <summary>
        /// Constructor for built in subsystem
        /// Defaults: lowQualityPixels = 5000, midQualityPixels = 10000, highQualityPixels = 15000
        /// lowQualityTime = 3s, midQyalityTime = 5s, highQualityTime = 7s
        /// </summary>
        /// <param name="EOSensorXmlNode"></param>
        /// <param name="dependencies"></param>
        /// <param name="asset"></param>
        public EOSensor(XmlNode EOSensorXmlNode, Dependency dependencies, Asset asset)
        {
            //DefaultSubName = "EOSensor";
            Asset = asset;
            GetSubNameFromXmlNode(EOSensorXmlNode);
            PIXELS_KEY = new StateVariableKey<double>(Asset.Name +"." + "numpixels");
            INCIDENCE_KEY = new StateVariableKey<double>(Asset.Name + "." + "incidenceangle");
            EOON_KEY = new StateVariableKey<bool>(Asset.Name + "." + "eosensoron");
            addKey(PIXELS_KEY);
            addKey(INCIDENCE_KEY);
            addKey(EOON_KEY);
            DependentSubsystems = new List<Subsystem>();
            SubsystemDependencyFunctions = new Dictionary<string, Delegate>();
            if (EOSensorXmlNode.Attributes["lowQualityPixels"] != null)
                //Console.WriteLine("inside loop");
                _lowQualityPixels = (double)Convert.ChangeType(EOSensorXmlNode.Attributes["lowQualityPixels"].Value.ToString(), typeof(double));
            if (EOSensorXmlNode.Attributes["lowQualityCaptureTime"] != null)
                _lowQualityCaptureTime = (double)Convert.ChangeType(EOSensorXmlNode.Attributes["lowQualityCaptureTime"].Value.ToString(), typeof(double));
            
            if (EOSensorXmlNode.Attributes["midQualityPixels"] != null)
                _midQualityPixels = (double)Convert.ChangeType(EOSensorXmlNode.Attributes["midQualityPixels"].Value.ToString(), typeof(double));
            if (EOSensorXmlNode.Attributes["midQualityCaptureTime"] != null)
                _midQualityCaptureTime = (double)Convert.ChangeType(EOSensorXmlNode.Attributes["midQualityCaptureTime"].Value.ToString(), typeof(double));
            
            if (EOSensorXmlNode.Attributes["highQualityPixels"] != null)
                _highQualityPixels = (double)Convert.ChangeType(EOSensorXmlNode.Attributes["highQualityPixels"].Value.ToString(), typeof(double));
            if (EOSensorXmlNode.Attributes["highQualityCaptureTime"] != null)
                _highQualityCaptureTime = (double)Convert.ChangeType(EOSensorXmlNode.Attributes["highQualityCaptureTime"].Value.ToString(), typeof(double));
            
            dependencies.Add("PowerfromEOSensor"+"."+Asset.Name, new Func<Event, HSFProfile<double>>(POWERSUB_PowerProfile_EOSENSORSUB));
            dependencies.Add("SSDRfromEOSensor" + "." + Asset.Name, new Func<Event, HSFProfile<double>>(SSDRSUB_NewDataProfile_EOSENSORSUB));
        }

        /// <summary>
        /// Constructor for scripted subsystem
        /// </summary>
        /// <param name="EOSensorXmlNode"></param>
        /// <param name="asset"></param>
        public EOSensor(XmlNode EOSensorXmlNode, Asset asset) : base(EOSensorXmlNode, asset)
        {
            
        }
        #endregion

        #region Methods
        /// <summary>
        /// An override of the Subsystem CanPerform method
        /// </summary>
        /// <param name="proposedEvent"></param>
        /// <param name="environment"></param>
        /// <returns></returns>
        public override bool CanPerform(Event proposedEvent, Domain environment)
        {
            //if (!base.CanPerform(proposedEvent, environment))
            //    return false;
            if (_task.Type == "imaging")
            {
                //set pixels and time to caputre based on target value
                int value = _task.Target.Value;
                double pixels = _lowQualityPixels;
                double timetocapture = _lowQualityCaptureTime;
                if (value <= 7 && value >= 5) //Morgan took out magic numbers
                {
                    pixels = _midQualityPixels;
                    timetocapture = 5;
                }
                if (value > 7)
                {
                    pixels = _highQualityPixels;
                    timetocapture = _highQualityCaptureTime;
                }

                // get event start and task start times
                double es = proposedEvent.GetEventStart(Asset);
                double ts = proposedEvent.GetTaskStart(Asset);
                double te = proposedEvent.GetTaskEnd(Asset);
                if (ts > te)
                {
                    // TODO: Change this to Logger
                    Console.WriteLine("EOSensor lost access");
                    return false;
                }

                // set task end based upon time to capture
                te = ts + timetocapture;
                proposedEvent.SetTaskEnd(Asset, te);

                // calculate incidence angle
                // from Brown, Pp. 99
                DynamicState position = Asset.AssetDynamicState;
                double timage = ts + timetocapture / 2;
                Matrix<double> m_SC_pos_at_tf_ECI = position.PositionECI(timage);
                Matrix<double> m_target_pos_at_tf_ECI = _task.Target.DynamicState.PositionECI(timage);
                Matrix<double> m_pv = m_target_pos_at_tf_ECI - m_SC_pos_at_tf_ECI;
                Matrix<double> pos_norm = -m_SC_pos_at_tf_ECI / Matrix<double>.Norm(-m_SC_pos_at_tf_ECI);
                Matrix<double> pv_norm = m_pv / Matrix<double>.Norm(m_pv);

                double incidenceang = 90 - 180 / Math.PI * Math.Acos(Matrix<double>.Dot(pos_norm, pv_norm));

                // set state data
                _newState.AddValue(INCIDENCE_KEY, timage, incidenceang);
                //_newState.AddValue(INCIDENCE_KEY, timage + 1, 0.0);

                _newState.AddValue(PIXELS_KEY, new KeyValuePair<double, double>(timage, pixels));
                _newState.AddValue(PIXELS_KEY, new KeyValuePair<double, double>(timage + 1, 0.0));

                _newState.AddValue(EOON_KEY, new KeyValuePair<double, bool>(ts, true));
                _newState.AddValue(EOON_KEY, new KeyValuePair<double, bool>(te, false));
            }
                return true;
            
        }

        /// <summary>
        /// Dependency Function for Power Subsystem
        /// </summary>
        /// <param name="currentEvent"></param>
        /// <returns></returns>
        public HSFProfile<double> POWERSUB_PowerProfile_EOSENSORSUB(Event currentEvent)
        {
            HSFProfile<double> prof1 = new HSFProfile<double>();
            prof1[currentEvent.GetEventStart(Asset)] = 10;
            if (currentEvent.State.GetValueAtTime(EOON_KEY, currentEvent.GetTaskStart(Asset)).Value)
            {
                prof1[currentEvent.GetTaskStart(Asset)] = 60;
                prof1[currentEvent.GetTaskEnd(Asset)] = 10;
            }
            return prof1;
        }

        /// <summary>
        /// Dependecy function for the SSDR subsystem
        /// </summary>
        /// <param name="state"></param>
        /// <returns></returns>
        public HSFProfile<double> SSDRSUB_NewDataProfile_EOSENSORSUB(Event currentEvent)
        {
            return currentEvent.State.GetProfile(PIXELS_KEY) / 500;
        }
        #endregion
    }
}
